### PR TITLE
Use the inner pattern to recreate the decision DAG for is-expressions

### PR DESCRIPTION
**Pull request by alrz**
Fixes https://github.com/dotnet/roslyn/issues/65876

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit2/Semantics/PatternMatchingTests_ListPatterns.cs [hunk 1]
```diff
@@ -3090,6 +3090,43 @@ public void Test2(int[] a)
...
+        var comp = CreateCompilationWithIndexAndRangeAndSpan(source, options: TestOptions.ReleaseExe);
+        CompileAndVerify(comp, expectedOutput: @"
+False
+False
+False
+False
+False
+False
+");
+    }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit2/Semantics/PatternMatchingTests_ListPatterns.cs:3128]: Consider adding similar tests for switch statement, switch expression and in a nested position in a pattern (basically other places where a negated pattern can go).

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1) with a test suggestion

File: src/Compilers/CSharp/Portable/BoundTree/BoundIsPatternExpression.cs [hunk 2]
```diff
@@ -13,11 +13,13 @@ public BoundDecisionDag GetDecisionDagForLowering(CSharpCompilation compilation)
...
             if (decisionDag.ContainsAnySynthesizedNodes())
             {
+                bool negated = this.Pattern.IsNegated(out var innerPattern);
+                Debug.Assert(negated == this.IsNegated);
                 decisionDag = DecisionDagBuilder.CreateDecisionDagForIsPattern(
                     compilation,
                     this.Syntax,
                     this.Expression,
-                    this.Pattern,
+                    innerPattern,
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/BoundIsPatternExpression.cs:22]: :memo: For other reviewers, we sometimes generate two DAGs: one for binding/analysis and one for lowering. See the condition above (`ContainsAnySynthesizedNodes()` check), which only triggers for list-patterns at the moment. 
The first DAG is produced by `Binder.MakeIsPatternExpression` and it has the proper `IsNegated` logic that was missing here.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundIsPatternExpression.cs:22]: >innerPattern [](http://example.com/codeflow?start=20&length=12)

It is not clear why stripping `not` here is the right thing to do. Is there some requirement that this API, ```GetDecisionDagForLowering```, supposed to do this? There is no hint about this on the API. Why nothing special is done for the code path that doesn't enter this ```if```. Perhaps the negation should be handled somewhere else?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/BoundTree/BoundIsPatternExpression.cs:22]: Ok, I see that we perform the same digging when we create ReachabilityDecisionDag during initial binding and here we should follow the same pattern.

**Comment by AlekseyTs** [comment]
@alrz Please consider using more meaningful titles for PRs. Often, simply describing the code change literally doesn't help the reader to reason about the nature of the change and the impact it has. For example, "inner pattern" is not some well known term. It is just a name of the local and parameter. So, to me "Use the inner pattern to recreate the decision DAG for is-expressions" sounds like a riddle that I have to solve. And I suspect that anyone running into a commit with a title like that would have to go through this too. 
Often, it helps to start with formulating what is the source of the undesired behavior. This is what we are eliminating/fixing and that is what could be good to include into the title. The how we achieve the goal could be important too, but it is secondary and could be included in the description.   
For this particular PR, I would say that the problem was the fact that, under some conditions,  BoundIsPatternExpression.GetDecisionDagForLowering violated the requirement that the decision dag represents the pattern after any outer 'not's are removed. This is what consumers expect. Essentially this is what I would try to put into the title. For example: "BoundIsPatternExpression.GetDecisionDagForLowering - when recreating a DAG, do that for the the pattern with any  'not's removed, as consumers expect." BTW, this is what I will use for the title when I merge this PR.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 2)

**Comment by AlekseyTs** [comment]
@alrz Thanks for the contribution.

**Pull request merged**